CorsixTH/Lua/app.lua

```diff
@@ -1680,60 +1680,136 @@ function App:loadLuaFolder(dir, no_results, append_to)
+  {major = 0, minor = 67, revision = 0, patch = "", version = 180},
+  {major = 0, minor = 68, revision = 0, patch = "", version = 194},
+  {major = 0, minor = 69, revision = 0, patch = "-beta-1", version = 216}
+}
+
+--! Retrieve the current savegame version as defined in the application.
+function App:getCurrentVersion()
+  return self.savegame_version
+end
+
+--! Requests data regarding a given savegame version
+--!param savegame_version (number) What to lookup, uses application version if blank
+--!return the matching release table entry
+-- If no releases match it returns the base release with the savegame version
+-- inserted
+function App:getReleaseData(savegame_version)
+  savegame_version = savegame_version or self:getCurrentVersion()
+  local release_data
+  for i = #release_table, 1, -1 do
+    local release = release_table[i]
+    if release.version == savegame_version then
+      release_data = release
+      break
+    elseif (release.version - savegame_version) < 0 then
+      -- we're not on a release version
+      local develop = shallow_clone(release) -- prevent recursion
+      develop.version = savegame_version
+      develop.patch = release.patch .. "-dev" .. savegame_version
+      release_data = develop
+      break
+    end
+  end
+  return release_data
+end
+
+--! Provides the release (or development) string for a given savegame version
+--!param savegame_version (number) The version to look up
+--!return A string in the format "v<major>.<minor>.<revision>[patch]"
+--! For development builds, the savegame version is appended as a patch e.g. 'dev213'
+function App:getReleaseString(savegame_version)
+  local release = self:getReleaseData(savegame_version)
+  local release_string = "v" .. release.major .. "." .. release.minor .. "." ..
+      release.revision
+  if string.len(release.patch) > 0 then
+    release_string = release_string .. release.patch
+  end
+  return release_string
+end
+
+--! Reports a difference between two versions based on requested methodology.
+--!param version_a (number or table) The first (usually newer) version to test
+--!param version_b (number or table) The second (usually older) version to test
+--!param method (string) What method to compare by
+--- method(release) reports the difference between the matching releases in steps,
+--- revisions are not counted.
+--- For development builds it will use the base release it started from when using
+--- the release method
+--- method(version) reports the difference between two savegame versions
+--!return The step difference between release a and release b for release method
+--- or The raw savegame version difference for version method
+function App:compareVersions(version_a, version_b, method)
+  assert((type(version_a) == "table" or type(version_a) == "number") and
+      (type(version_b) == "table" or type(version_b) == "number"),
+      "Compare requires savegame version or the version table to function")
```

Comment:
```suggestion
      "Compare requires savegame version or an entry from the version table to function")
```
Likely you don't want the entire table as parameter :)

If you split the check (one check for parameter `version_a` and one check for parameter `version_b`), it's simpler to understand what is wrong precisely.
Alternatively, add the values to the error string, so we can read what's broken.

Reply:
Whole assert edited above, had to amend your suggestion just so it preserved your co-author